File: src/Tree/ConstructBinaryTreefromPreorderandInorderTraversal/ConstructBinaryTreefromPreorderandInorderTraversal.java
class Solution {
    Map<Integer, Integer> map = new HashMap<>();
    int index = 0;
    int[] preorder_arr;
    int[] inorder_arr;
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        //Define variables
        preorder_arr = preorder;
        inorder_arr = inorder;
        //Define the table
        for(int i = 0; i < inorder.length; i++){
            map.put(inorder[i], i);
        }
        //Pass the boundaries
        return dfs(0, inorder.length - 1);
    }
    private TreeNode dfs(int left, int right){
        int curVal = preorder_arr[index++];
        int curIdx = map.get(curVal);
        //Build current node
        TreeNode root = new TreeNode(curVal);
            
        //Build left subtree
        if(left < curIdx){
            root.left = dfs(left, curIdx - 1);
        }
        //Build right subtree
        if(curIdx < right){
            root.right = dfs(curIdx + 1, right);
        }
        //Return current root node
        return root;
    }
}

File: src/Tree/ConstructBinaryTreefromPreorderandInorderTraversal/ConstructBinaryTreefromPreorderandInorderTraversal.py
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        # 1. Save each inorder element's value and index onto map
        map = dict()

        for i in range(len(inorder)):
            map[inorder[i]] = i

        # 2. perform dfs to build tree
        idx = 0

        def dfs(start: int, end: int):
            # base check
            if end < start:
                return None
            
            nonlocal idx
            root_val = preorder[idx]
            idx += 1

            inorder_index = map[root_val]

            left_node = dfs(start, inorder_index - 1)
            right_node = dfs(inorder_index + 1, end)

            root = TreeNode(val=root_val, left=left_node, right=right_node)

            return root

        return dfs(0, len(inorder) - 1)
